data_structures/heap/heap.py

```diff
@@ -1,99 +1,131 @@
-#!/usr/bin/python3
+from typing import Iterable, List, Union
 
 
 class Heap:
-    """
+    """A Max Heap Implementation
+
     >>> unsorted = [103, 9, 1, 7, 11, 15, 25, 201, 209, 107, 5]
     >>> h = Heap()
-    >>> h.build_heap(unsorted)
-    >>> h.display()
+    >>> h.build_max_heap(unsorted)
+    >>> print(h)
     [209, 201, 25, 103, 107, 15, 1, 9, 7, 11, 5]
     >>>
-    >>> h.get_max()
+    >>> h.extract_max()
     209
-    >>> h.display()
+    >>> print(h)
     [201, 107, 25, 103, 11, 15, 1, 9, 7, 5]
     >>>
     >>> h.insert(100)
-    >>> h.display()
+    >>> print(h)
     [201, 107, 25, 103, 100, 15, 1, 9, 7, 5, 11]
     >>>
     >>> h.heap_sort()
-    >>> h.display()
+    >>> print(h)
     [1, 5, 7, 9, 11, 15, 25, 100, 103, 107, 201]
-    >>>
     """
 
     def __init__(self):
-        self.h = []
-        self.curr_size = 0
+        self.h: List[Union[int, float]] = []
+        self.heap_size: int = 0
+
+    def __repr__(self) -> str:
+        return str(self.h)
+
+    def parent_index(self, child_idx: int) -> Union[int, None]:
+        """ return the parent index of given child """
+        if child_idx > 0:
+            return (child_idx - 1) // 2
+        return None
 
-    def get_left_child_index(self, i):
-        left_child_index = 2 * i + 1
-        if left_child_index < self.curr_size:
+    def left_child_idx(self, parent_idx: int) -> Union[int, None]:
+        """
+        return the left child index if the left child exists.
+        if not, return None.
+        """
+        left_child_index = 2 * parent_idx + 1
+        if left_child_index < self.heap_size:
             return left_child_index
         return None
 
-    def get_right_child(self, i):
-        right_child_index = 2 * i + 2
-        if right_child_index < self.curr_size:
+    def right_child_idx(self, parent_idx: int) -> Union[int, None]:
+        """
+        return the right child index if the right child exists.
+        if not, return None.
+        """
+        right_child_index = 2 * parent_idx + 2
+        if right_child_index < self.heap_size:
             return right_child_index
         return None
 
-    def max_heapify(self, index):
-        if index < self.curr_size:
-            largest = index
-            lc = self.get_left_child_index(index)
-            rc = self.get_right_child(index)
-            if lc is not None and self.h[lc] > self.h[largest]:
-                largest = lc
-            if rc is not None and self.h[rc] > self.h[largest]:
-                largest = rc
-            if largest != index:
-                self.h[largest], self.h[index] = self.h[index], self.h[largest]
-                self.max_heapify(largest)
-
-    def build_heap(self, collection):
-        self.curr_size = len(collection)
+    def max_heapify(self, index: int):
```

Comment:
This function does not have a return statement which means it implicitly returns `None`.
```suggestion
    def max_heapify(self, index: int) -> None:
```